core/src/main/java/org/apache/accumulo/core/constraints/DefaultKeySizeConstraint.java

```diff
@@ -53,16 +53,14 @@ public String getViolationDescription(short violationCode) {
       return NO_VIOLATIONS;
 
     List<Short> violations = new ArrayList<>();
-
+    int size = mutation.getRow().length;
     for (ColumnUpdate cu : mutation.getUpdates()) {
-      int size = mutation.getRow().length;
```

Comment:
The limited documentation on the class says it limits the size of keys.  A column update would correspond to a key, so the existing code may correct.  One change would be that it could break out of the loop after adding the first violation.

Reply:
I wasn't sure but to me the `row ID` is an integral part of the Key and a user could very easily create a key larger than the constraint if they have a large `row ID`

Reply:
> a user could very easily create a key larger than the constraint if they have a large row ID

The row length was being factored into the size of each column update which would equate to the eventual key size.

Reply:
Yes but if there aren't any Column updates the row length will be ignored. Adding this chunk of code to the test will cause it to fail now:
<pre>
m = new Mutation(oversized);
assertEquals(Collections.singletonList(DefaultKeySizeConstraint.MAX__KEY_SIZE_EXCEEDED_VIOLATION), 
        constraint.check(null, m));
</pre>

Reply:
The `BatchWriter` at least prevents this from every happening and will throw an exception:
https://github.com/apache/accumulo/blob/675e41aaed0f8ae250d97a10dad35a8aef4c3e82/core/src/main/java/org/apache/accumulo/core/clientImpl/TabletServerBatchWriter.java#L247

Reply:
The current code seems to impose a limit on any total key size (4 dimensions: row + cf + cq + cv; doesn't include timestamp or internal delete flag dimensions). Your change seems to impose a limit on the total size of this mutation, and seems to turn it into a MutationSizeConstraint.